### PR TITLE
Refactor inversion inputs reimplement

### PR DESCRIPTION
**Pull request by brendan-m-murphy**
* **Summary of changes**

A previous PR #355 extracted a function `make_inv_inputs` from `fixedbasisMCMC` to create the arrays needed for PyMC from the `fp_data` dictionary. This PR re-implements `make_inv_inputs` using newer methods that leverage xarray.

I also removed arguments from `make_inv_inputs` that were just "passed through", so now the function returns two dictionaries of np.ndarrays: one for the mcmc args, and one for post-processing.

Tests have been added to ensure the new functions keep the old behaviour.

For `monthly_bcs`, there were some edge cases that weren't previously tested:
- if the end date was in the middle of the month, there would be no BC period for that month, so there would be rows of all zeros in `H_bc`
- if the start date was before any actual values in the BC data, then there would be BC periods corresponding to times without data, so there would be columns of all zeros in `H_bc`.

The new tests check this (by xfail tests for assertions that there are not rows or columns of zeros). The new method for creating monthly BCs passes this check.

Note: eventually, it would be good to update the PyMC code to accept xarray inputs directly. Currently we pass numpy arrays and I reconstruct the coordinates to give to PyMC. That would be a separate change though, since I don't want this PR to grow in scope.

* **Please check if the PR fulfills these requirements**

- [ ] Closes #xxxx (Replace xxxx with the Github issue number)
- [x] Tests added and passing
- [ ] Documentation and tutorials updated/added
- [x] Added an entry to the `CHANGELOG.md` file
- [ ] Added any new requirements to `requirements.txt`

### CODE REVIEW

File: openghg_inversions/hbmcmc/hbmcmc.py [hunk 1]
```diff
@@ -102,148 +112,55 @@ def make_inv_inputs(
...
     post_process_args = {
-        "Ytime": Ytime,
+        "Ytime": y_time,
         "obs_repeatability": obs_repeatability,
         "obs_variability": obs_variability,
-        "obs_prior_factor": obs_prior_factor,
-        "obs_prior_upper_level_factor": obs_prior_upper_level_factor,
+        "obs_prior_factor": (
+            obs_prior_factor if obs_prior_factor is not None else np.zeros_like(y)
+        ),  # use zeros instead of None
```
[the remark this answers is not in the document]
Comment (brendan-m-murphy) [reply openghg_inversions/hbmcmc/hbmcmc.py:159]: Yes I caught this an pushed an update after I asked for the review.

File: openghg_inversions/inversion_inputs.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    # Just assigning ds["H_bc"] = temp is fine, but emits a warning.
+    temp_values_only = xr.DataArray(
+        temp.data,
+        dims=("bc_region", "nmeasure"),
+        coords={"bc_region": temp["bc_region"]},  # keep only the new dim coord(s)
+        name="H_bc",
+    )
+
+    ds["H_bc"] = temp_values_only
+    return ds
```
[the remark this answers is not in the document]
Comment (brendan-m-murphy) [reply openghg_inversions/inversion_inputs.py:284]: I'm not totally sure what is right here, but `coords={"bc_region": ..., "nmeasure": ds["nmeasure"]}` will trigger a warning from pandas (via xarray).

File: openghg_inversions/inversion_inputs.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+        # record times
+        ragged_vals.append(v[ragged_dim])
+
+    # concat pieces
+    da = xr.concat(pieces, dim=stack_dim, **concat_kwargs)
+
+    # now create and assign multi-index
+    key_indicator = np.concat(key_vals)
+    concat_ragged = np.concat(ragged_vals)
```
[the remark this answers is not in the document]
Comment (brendan-m-murphy) [reply openghg_inversions/inversion_inputs.py]: np.concat exists (there is an alias in the numpy code somewhere), but will switch to the canonical np.concatenate.

File: openghg_inversions/inversion_inputs.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        _, inv = np.unique(arr, return_inverse=True)
+        return inv
+
+    return xr.apply_ufunc(np_unique_inv, da)
+
+
+# MAKE FUNCTIONS
+def make_site_indicator(site_coord: xr.DataArray) -> xr.DataArray:
+    return xr_unique_inv(site_coord).rename("site_indicator")
+
```
[the remark this answers is not in the document]
Comment (brendan-m-murphy) [reply openghg_inversions/inversion_inputs.py:179]: Nice catch, I added a sort option that uses pd.factorize when false.

File: openghg_inversions/inversion_inputs.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+    ds: xr.Dataset,
+    fp_data: dict[str, Any],
+    min_error: str | dict[str, float] | float = 0.0,
+    min_error_per_site: bool = True,
+) -> xr.Dataset:
+    """Add min_error to combined Dataset."""
+    if isinstance(min_error, float) or (isinstance(min_error, np.ndarray) and min_error.ndim == 0):
+        ds["min_error"] = min_error * xr.ones_like(ds.mf)
+    elif isinstance(min_error, dict):
+        sites = [k for k in fp_data if not k.startswith(".")]
```
[the remark this answers is not in the document]
Comment (brendan-m-murphy) [reply openghg_inversions/inversion_inputs.py:222]: I don't think the dictionary option is actually exposed to the users... but this would be something to check if it were. (I think this might have been an option I used when experimenting.)

File: openghg_inversions/inversion_inputs.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    """
+    dvs = next(iter(ds_dict.values())).data_vars
+
+    gathered_dvs = {}
+
+    for dv in dvs:
+        da_dict = {k: v[dv] for k, v in ds_dict.items()}
+        gathered_dvs[dv] = concat_gather_data_arrays(da_dict, key_dim, ragged_dim, stack_dim, **concat_kwargs)
+
+    return xr.Dataset(gathered_dvs)
```
[the remark this answers is not in the document]
Comment (brendan-m-murphy) [reply openghg_inversions/inversion_inputs.py:103]: I added a check that raises a value error. It would be nice to extend this to work with missing data variables (which might happen in an inversion with both satellite and in-situ obs), but maybe that should be a future enhancement.

File: tests/test_inversion_inputs.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+@pytest.mark.create_frozen
+def test_inversion_input_create_frozen(raw_data_path, inv_inputs_args):
+    """This 'test' just regenerates frozen data for use in other tests."""
+    mcmc_args, post_args = make_inv_inputs(**inv_inputs_args)
+
+    out_name = raw_data_path / "frozen_mhd_tac_make_inv_inputs_hbmcmc.npz"
+    save_frozen_npz(out_name, mcmc_args=mcmc_args, post_process_args=post_args)
+
+
+def test_inversion_input_hbmcmc_matches_frozen(raw_data_path, inv_inputs_args):
```
Comment (ben-adam63) [inline tests/test_inversion_inputs.py:101]: This test is failing with me, with very very small differences between the arrays (differences on the order 10^-13, so negligible compared to the real array). Maybe a `rtol` or `atol` in the `np.assert_allclose` would be useful?

**Review by qq23840 — COMMENTED** [review]
One issue about the test that I've raised, but otherwise looks good to me - my usual test inversions seem to run just fine.

File: tests/test_inversion_inputs.py (hunk 7 above)
Comment (brendan-m-murphy) [reply tests/test_inversion_inputs.py:101]: Okay I added a small tolerance. The test data can be "regenerated" by running `pytest -m create_frozen`, but I didn't do this by default since it's a bit wasteful to do every time the tests run. The tolerance should help...

**Review by SutarPrasad — APPROVED** [review]
